src/functions/authoriser/application/__tests__/AdJwtVerifier.spec.ts

```diff
@@ -0,0 +1,106 @@
+import { Mock, It, Times } from 'typemoq';
+import * as jwt from 'jsonwebtoken';
+import { AdJwtVerifier, JwksClient, JsonWebKey } from '../AdJwtVerifier';
+
+describe('AdJwtVerifier', () => {
+  const moqJwtDecode = Mock.ofInstance(jwt.decode);
+  const moqJwtVerify = Mock.ofInstance(jwt.verify);
+  const moqJwksClient = Mock.ofType<JwksClient>();
+
+  let testSigningKey: JsonWebKey;
+
+  let sut: AdJwtVerifier;
+
+  beforeEach(() => {
+    moqJwtDecode.reset();
+    moqJwtVerify.reset();
+    moqJwksClient.reset();
+
+    moqJwtDecode.setup(x => x(It.isAnyString(), It.isAny()))
+      .returns(() => ({ header: { typ: 'test-type', alg: 'test-alg', kid: 'test-kid' } }));
+
+    moqJwksClient.setup(x => x.getSigningKey(It.isAnyString()))
+      .returns(() => Promise.resolve(testSigningKey));
+
+    moqJwtVerify.setup(x => x(It.isAnyString(), It.isAnyString(), It.isAny()))
+      .returns(() => ({ sub: 'test-subject', unique_name: 'test-unique_name' }));
+
+    spyOn(jwt, 'decode').and.callFake(moqJwtDecode.object);
```

Comment:
Interesting, what's the benefit of setting up your mocks via typemoq vs doing it all natively in Jasmine like below?

```
spyOn(<object>, '<functionName>').and.returnValue(<stubValue>)'
```

Reply:
@johnmorleybjss The benefits include:

- Type safety.  The stub value using Jasmine spies is an any, whereas the stub value using TypeMoq is the actual type of the return value.  If the type changes, for example new properties get added, or the entire type changes, then we'd get compilation errors with TypeMoq.
- Similar to type safety, we get better discoverability in IDEs, because it will hint to us the expected type, and also warn us if we've got it wrong.
- It allows for more loose expectation verifications, allowing for less brittle tests.  For example, when verifying that an error was logged, we can use: `moqConsoleLog.verify(x => x(It.is<string>(s => /Failed authorization, responding with Deny/.test(s))), Times.once());`.  There's lots more information in the logging message, but that's not important to the test.  This means we can change the additional information in the logs, without breaking the test.  I know Jasmine can test for specific arguments, but only whole values not predicates tests or regular expressions.
- Enables an easier to understand layout to tests files, shared initialisation which is exactly the same can go in a single `beforeEach` function, but then the chaining to setup the specific scenario can go in each test.  This is kind of possible using Jasmine spies, but the code is less easy to read.

I realise that TypeMoq does introduce another library, and that the setup code is more verbose because we have to specific the arguments using the `It` helper, but that is what allows for more finer control.  However, if you'd prefer to not use TypeMoq, I'm happy to take it out and change to just using Jasmine spies.

Its a shame that TypeMoq doesn't provide functionality to actually stub out functions in place, like spies can do, because then we wouldn't have to mix TypeMoq and spies.

Reply:
I think we need to take some time to understand these benefits and decide on whether we want to adopt this style across the board. Happy to merge either way for now.